Fix edge case for tokenize (#36277)

wangzhen0518 opened the pull request:
# What does this PR do?

Fixes #36277

The dtype of the tensor returned by tokenize is inconsistent depending on whether the input text is empty or not, e.g.,
```python
def show_return_dtype(x:str):
    return_type = ["tf", "pt", "jax", "mlx", "np"]
    for t in return_type:
        output = tokenizer(x, return_tensors=t)
        print(f"{t}: {output['input_ids'].dtype}")

x = ""
show_return_dtype(x)
# tf: <dtype: 'float32'>
# pt: torch.float32
# jax: float32
# mlx: mlx.core.float32
# np: float64

x = "abc"
show_return_dtype(x)
# tf: <dtype: 'int32'>
# pt: torch.int64
# jax: int32
# mlx: mlx.core.int32
# np: int64
```

After this PR is fixed:
```python
def show_return_dtype(x:str):
    return_type = ["tf", "pt", "jax", "mlx", "np"]
    for t in return_type:
        output = tokenizer(x, return_tensors=t)
        print(f"{t}: {output['input_ids'].dtype}")

x = ""
show_return_dtype(x)
# tf: <dtype: 'int32'>
# pt: torch.int64
# jax: int32
# mlx: mlx.core.int32
# np: int64

x = "abc"
show_return_dtype(x)
# tf: <dtype: 'int32'>
# pt: torch.int64
# jax: int32
# mlx: mlx.core.int32
# np: int64
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
      https://github.com/huggingface/transformers/issues/36277
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@Rocketknight1 and @ArthurZucker

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts, @qubvel
- speech models: @ylacombe, @eustlb
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Rocketknight1
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @SunMarc
- chat templates: @Rocketknight1

Integrations:

- deepspeed: HF Trainer/Accelerate: @muellerzr
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc @MekkCyber

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

wangzhen0518 committed:
Fix edge case for tokenize (#36277)

Rocketknight1 commented:
cc @arthurzucker @itazap for review. The full explanation is at #36277, but the tl;dr is:

- Tokenizer tensor outputs are generated with `torch.tensor()`
- `torch.tensor()` infers the dtype according to the input data and the following rule:
  - If every element is an int, `dtype=torch.int64`
  - If at least one element is a float, `dtype=torch.float32`
  - If **the input list is empty**, `dtype=torch.float32`

This creates an edge cases when our tokenizers return empty arrays - the dtype switches from `torch.int64` to `torch.float32`, and this breaks ops like `torch.concat`. This PR **always** casts the dtype to `torch.int64`, but is there any case where tokenizers return float32 arrays? I can't think of one, but this will definitely break those cases if they exist!

Rocketknight1 commented:
Rebasing to run tests, gentle ping @arthurzucker @itazap

wangzhen0518 committed:
Fix tokenizing dtype for float input cases

wangzhen0518 commented:
It seems that some testcases in speech models have float tokenizing inputs. I modified the code to handle this case and fixed a bug in the previous commit. Now, the dtype is set to int only when the value is empty and the user has not set the dtype. In all other cases, it remains unchanged from before this PR.

wangzhen0518 commented:
I don't understand the two failed tests. They appear to be an issue with the model files on Hugging Face, unrelated to tokenization.

> FAILED tests/models/wav2vec2/test_modeling_wav2vec2.py::Wav2Vec2ModelTest::test_pipeline_audio_classification - FileNotFoundError: Couldn't find any data file at /root/project/hf-internal-testing/librispeech_asr_dummy. Couldn't find 'hf-internal-testing/librispeech_asr_dummy' on the Hugging Face Hub either: LocalEntryNotFoundError: An error happened while trying to locate the file on the Hub and we cannot find the requested files in the local cache. Please check your connection and try again or make sure your Internet connection is on.
>
> FAILED tests/models/wav2vec2/test_modeling_wav2vec2.py::Wav2Vec2ModelTest::test_pipeline_audio_classification_fp16 - FileNotFoundError: Couldn't find any data file at /root/project/hf-internal-testing/librispeech_asr_dummy. Couldn't find 'hf-internal-testing/librispeech_asr_dummy' on the Hugging Face Hub either: LocalEntryNotFoundError: An error happened while trying to locate the file on the Hub and we cannot find the requested files in the local cache. Please check your connection and try again or make sure your Internet connection is on.

Rocketknight1 commented:
Yes, these are CI issues. You can ignore them, we'll try to get them fixed!

wangzhen0518 commented:
Hi, I just merged the latest main branch into this PR and all tests have passed. Are there any other issues need to be addressed?

itazap commented:
@wangzhen0518 Thanks for addressing this! 😊 Do you mind please sharing a code snippet / description of the edge case where the inconsistent typing was an issue, so that we can add a test for it? That way your change won't be undone in the future! 😉

wangzhen0518 commented:
# Inconsistent cases

When the input text is empty, the output tensor is an empty tensor of type float, such as `torch.float32`. When the input text is not empty, the output tensor is of type int, such as `torch.int64`. The types of the output tensors are inconsistent in these two cases.

```python
t = tokenizer('', return_tensors='pt')
print(t['input_ids'].dtype) # torch.float32

t = tokenizer('abc', return_tensors='pt')
print(t['input_ids'].dtype) # torch.int64
```

# The case of an empty input string is used when...
I use a LLM as an agent to interact with an environment, and the feedback from the environment sometimes is empty.
Here is an example. The `next_app_actor_prompt` sometimes is empty.
```python
actor_prompt = self.env.reset(query_config, args.seed, args.max_rollout_step, args.prompt_max_len)
inputs = self.tokenizer(
    [actor_prompt],
    return_tensors="pt",
    add_special_tokens=False,
    max_length=max_length,
    padding=True,
    truncation=True,
)
while not (failed or truncated or done):
    input_len: int = inputs["input_ids"].size(1)
    sequences, attention_mask, action_mask = self.actor.generate(**inputs, **generate_kwargs)
    action = self.get_action(sequences, input_len)
    # `next_app_actor_prompt` sometimes is an empty string.
    next_app_actor_prompt, reward, failed, truncated, done, _ = self.env.step(action)
    next_app_input_token = self.tokenize_fn(next_app_actor_prompt, self.prompt_max_len, device="cuda")
    next_inputs = {
        "input_ids": torch.hstack((sequences, next_app_input_token["input_ids"])),
        "attention_mask": torch.hstack((attention_mask, next_app_input_token["attention_mask"])),
    }
    inputs = next_inputs
    ...
```

# What I do in this PR
I changed the output type to int when the user does not specify the output type and the input is empty.

# Testcase
Here is a testcase demo.
```python
def test_empty_string():
    tokenizer = ...
    x = ""
    tokenizer_return_type = ["tf", "pt", "jax", "mlx", "np"]
    output_tensor_type = [tf.int32, torch.int64, jnp.int32, mx.int32, np.int64]
    for t1, t2 in zip(tokenizer_return_type, output_tensor_type):
        output = tokenizer(x, return_tensors=t1)
        assert output["input_ids"].dtype is t2
```

# Potential Issues
I noticed that the tokenizer processes audio data in some test cases. I'm not sure whether there could be cases where the input audio data is empty, and if so, whether the return type should be set to `int` when the audio input is empty (by default, it returns `float` when the audio is non-empty).

ArthurZucker reviewed (approved):
LGTM let's just add a test please!

wangzhen0518 committed:
add test for empty input string

wangzhen0518 committed:
deal empty list of list like [[]]

wangzhen0518 commented:
I added a test case, but I found that some tokenizers do not support tokenizing `str`. They require other input data structures. The unsupported tokenizers are as follows:  

- `LayoutLMv2Tokenizer`, `LayoutLMv3Tokenizer`, `LayoutXLMTokenizer`, and `UdopTokenizer` require input as `list[str]` along with `boxes`.  
- `MarkupLMTokenizer` requires input as `list[str]` along with `xpaths`.  
- `TapasTokenizer` requires input as `pd.DataFrame`.  

I'm not familiar with these types of tokenizers. Currently, I added the test case in `tests/test_tokenization_common.py`. Do I need to add separate test cases for these tokenizers?

Rocketknight1 commented:
I'm going to be away for a couple of weeks, so gentle ping @itazap @ArthurZucker just in case this PR goes missing. I've seen this issue reported elsewhere, e.g. https://github.com/huggingface/transformers/issues/38417#issuecomment-3012105880

itazap commented:
hello @wangzhen0518  thanks for adding the test! 😊 Yes those models handle different types of data and so they often need custom versions of tests that override the common test. If you take a look at their test files, they often just require additional inputs such as `boxes` for `layout` models, etc. Usually the override tests is simply the original test + the necessary additional arg like `boxes` (can be copied from other tests)

wangzhen0518 committed:
add tests for tokenizer for models with input that is not plain text

wangzhen0518 commented:
I have added tests for `LayoutLMv2Tokenizer`, `LayoutLMv3Tokenizer`, `LayoutXLMTokenizer`,`UdopTokenizer`, `MarkupLMTokenizer`, and `TapasTokenizer`. I'm not sure what the github-actions above means—should I mark these tests as slow?

ArthurZucker commented:
No don't worry we just ran the full slow CI so should be alright

ArthurZucker reviewed (approved):
Can we just simplify the tests to just the updated common tests please? then we can merge!

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_36555). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

wangzhen0518 commented:
> Can we just simplify the tests to just the updated common tests please? then we can merge!

Since the input parameters are different, `LayoutLMv2Tokenizer`, `LayoutLMv3Tokenizer`, `LayoutXLMTokenizer`, `UdopTokenizer`, `MarkupLMTokenizer`, and `TapasTokenizer` require separate tests. Do you mean to only keep the common tests in `tests/test_tokenization_common.py` and set the tests for the above tokenizers to `skip`?
